sources/packages/backend/apps/api/src/services/restriction/student-restriction.service.ts

```diff
@@ -406,7 +426,99 @@ export class StudentRestrictionService extends RecordDataModelService<StudentRes
         auditUserId,
         applicationId,
       );
-      await entityManager.getRepository(StudentRestriction).save(restriction);
+      const newRestriction = await entityManager
+        .getRepository(StudentRestriction)
+        .save(restriction);
+      await this.createNotifications(
+        [newRestriction.id],
+        auditUserId,
+        entityManager,
+      );
     }
   }
+
+  /**
+   * Generate notifications for newly created student restrictions when needed.
+   * @param restrictionsIds ids to generate the notifications.
+   * @param auditUserId user that should be considered the one that is causing the changes.
+   * @param entityManager optional repository that can be provided, for instance,
+   * to execute the command as part of an existing transaction. If not provided
+   * the local repository will be used instead.
+   */
+  async createNotifications(
+    restrictionsIds: number[],
+    auditUserId: number,
+    entityManager?: EntityManager,
+  ): Promise<void> {
+    const restrictions = await this.getRestrictionsForNotifications(
+      restrictionsIds,
+      entityManager,
+    );
+
+    if (!restrictions?.length) {
+      // There are no notifications to be sent.
+      return;
+    }
+
+    const notifications = restrictions.map((restriction) => ({
+      givenNames: restriction.student.user.firstName,
+      lastName: restriction.student.user.lastName,
+      toAddress: restriction.student.user.email,
+      userId: restriction.student.user.id,
+    }));
+    await this.gcNotifyActionsService.sendStudentRestrictionAddedNotification(
+      notifications,
+      auditUserId,
+    );
+  }
+
+  /**
+   * Gets the notifications information for student restrictions that are
+   * eligible to generate a notification (actionType not defined as NoEffect).
+   * @param restrictionIds student restriction ids.
+   * @param entityManager optional repository that can be provided, for instance,
+   * to execute the command as part of an existing transaction. If not provided
+   * the local repository will be used instead.
+   * @returns student restrictions eligible to generate a notification.
+   */
+  private async getRestrictionsForNotifications(
+    restrictionIds: number[],
+    entityManager?: EntityManager,
+  ): Promise<StudentRestriction[]> {
+    const repository =
+      entityManager?.getRepository(StudentRestriction) ?? this.repo;
+    const allRestrictions: StudentRestriction[] = [];
+    while (restrictionIds.length > 0) {
+      // Breaks the execution in chunks to allow the selects of a huge amount of records.
```

Comment:
I didn't understand this.

Reply:
Hope that it is better after our chat. The splice will remove the items from the arrays and return the deleted ones to be processed till there are not items in the array anymore.